src/bia_tischi/_utilities.py

```diff
@@ -57,16 +57,79 @@ def find_variable(variables, key, type_checker_function):
     return variables[other_name]
 
 
-def generate_response(user_input: str, model):
+def generate_response_to_user(model, user_prompt: str):
     """Generates code and text respond for a specific user input.
     To do so, it combines the user input with additional context such as
     current variables and a prompt template."""
 
+    system_prompt = create_system_prompt()
+
+    # take the last n chat entries
     from ._machinery import Context
+    n = 10
+    chat_history = Context.chat[-n:]
+
+    if Context.verbose:
+        print("\nUser prompt:", user_prompt)
+        print("\nSystem prompt:", system_prompt)
+        print_chat(chat_history)
+
+    full_response = generate_response_from_openai(model, system_prompt, user_prompt, chat_history)
+
+    if Context.verbose:
+        print("\n\nFull response:\n", full_response)
+
+    pricing_summary = create_pricing_summary(model, chat_history, full_response, system_prompt, user_prompt)
+
+    # Search for the code pattern in the text
+    import re
+    pattern = re.compile(r'([\s\S]*?)```python([\s\S]*?)```')
+    match = pattern.search(full_response)
+    if match:
+        text = match.group(1).strip()
+        code = match.group(2).strip()
+    else:
+        text = full_response
+        code = None
+
+    text = pricing_summary + "### Assistant response\n\n" + text
+    text += "\n#### Additional information\n\n"
 
+    return code, text
+
+
+def create_pricing_summary(model, chat_history, full_response, system_prompt, user_prompt):
+    import tiktoken
+    from ._machinery import Models, Context
+    encoding = tiktoken.encoding_for_model(model)
+    input_token = (len(encoding.encode(user_prompt))
+                   + len(encoding.encode(system_prompt))
+                   + len(encoding.encode(concatenate_chat_content(chat_history))))
+    output_token = len(encoding.encode(full_response))
+    pricing = "\n##### Pricing\n\n"
+    pricing += "- Model: " + model + "\n"
+    pricing += "- Pricing: https://openai.com/pricing\n"
+    pricing += "- Last input: " + str(input_token) + " token = "
+    input_price = Models.usd_per_1k_input_token(model) * input_token / 10.0
+    pricing += print_costs_in_cent(input_price)
+    pricing += "- Last output: " + str(output_token) + " token = "
+    output_price = Models.usd_per_1k_output_token(model) * output_token / 10.0
+    pricing += print_costs_in_cent(output_price)
+    Context.session_price_us_cent += input_price + output_price
+    pricing += "- Total session costs: " + print_costs_in_cent(Context.session_price_us_cent)
+    pricing += "\n"
+    return pricing
```

Comment:
Out of curiosity: is this related to implementing _memory_? I would leave the pricing thing out for now as it might be a research topic itself.

Reply:
no it is not, I did bunch of stuff in parallel